proxy/http.go

```diff
@@ -0,0 +1,196 @@
+		// might be meant to be interpreted by the connector itself.
+		if !isPublicPort {
+			var claims, authErr = auth.AuthorizedReq(req, h.jwtVerificationKey)
+			if authErr == nil {
+				authErr = auth.EnforcePrefix(claims, proxyConn.taskName)
+			}
+			if authErr != nil {
+				handleHttpError(authErr, w, req)
+				return
+			}
+		}
+		proxy.ServeHTTP(w, req)
+	})
+
+	if clientConn.ConnectionState().NegotiatedProtocol == "h2" {
+		var h2Server = http2.Server{}
+		// The clientConn will be closed automatically by ServeConn, but we'll need to close the proxyConn ourselves
+		h2Server.ServeConn(clientConn, &http2.ServeConnOpts{
+			Context: ctx,
+			Handler: handlerFunc,
+		})
+		// TODO: h2Server.ServeConn does not return an error, so not really sure what to return here
+		return nil
+	} else {
+		// We'll be speaking http/1.1, which requires a 3rd party library because there's no ServeConn function in Go's http package.
+		var server = fasthttp.Server{
+			Handler: fasthttpadaptor.NewFastHTTPHandler(handlerFunc),
+			Name:    fmt.Sprintf("%s (%s)", proxyConn.hostname, clientConn.RemoteAddr().String()),
+		}
+
+		return server.ServeConn(clientConn)
+	}
+}
+
+var errTemplate = template.Must(template.New("proxy-error").Parse(`<!DOCTYPE html>
+<html>
+	<head>
+	    <title>Error</title>
+		<style>
+			html {
+				height: 100%;
+				display: table;
+				margin: auto;
+			}
+			body {
+				height: 100%;
+				display: table-cell;
+				vertical-align: middle;
+				background-color: white;
+			}
+		</style>
+	</head>
+	<body>
+		<span style='font-size: 40px; color: black; font-family:Arial,Helvetica,sans-serif;'>{{.}}</span>
+	</body>
+</html>`))
+
+func handleHttpError(err error, w http.ResponseWriter, r *http.Request) {
+	var body []byte
+	var contentType string
+	var status = httpStatus(err)
+
+	// If the error was caused by an issue with the upstream connection, then we must
+	// ask the client to close the connection and create a new one. This is important
+	// because the http proxy handler isn't currently able to re-establish connections
+	// in response to them breaking. So if the client continues to use this connection,
+	// then they will continue to get 5xx errors due to the broken upstream connection.
+	// Note that, while the `Connection` header is not compatible with http2, the Go
+	// http2 package seems to handle this by removing the header and sending a GOAWAY.
+	// In fact, this is the _only_ way I can figure out how to send a GOAWAY from a handler.
+	if status >= 500 {
+		w.Header().Add("Connection", "close")
+	}
+
+	var headers = r.Header["Accept"]
+	var accept string
+	if len(headers) > 0 {
+		accept = headers[0]
+	}
+	if strings.Contains(accept, "json") {
+		body, _ = json.Marshal(map[string]interface{}{
+			"error": err.Error(),
+		})
+		contentType = "application/json"
+	} else if strings.Contains(accept, "html") {
+		var buf bytes.Buffer
+		if templateErr := errTemplate.Execute(&buf, err.Error()); templateErr != nil {
+			// TODO:
```

Comment:
?

Reply:
:facepalm: that was supposed to be to remove the `panic` below :upside_down_face: